Initial Release 1.0.0

f3rno opened the pull request:
Description pending, see readme

### TODO
- [x] add api credential storage/loading (currently reads `.env` in load dir)
- [x] support STOP LIMIT/TRAILING STOP
- [x] finalize for release
- [x] fix settings file directory

### For Review
After the code review, please verify each individual command works as expected. The full list can be viewed with `--help`:
```
Usage: bfx-cli [command] <options>

Commands:
  bfx-cli affiliate <code>                  Set affiliate code
  bfx-cli credentials                       Configure API credentials
  bfx-cli watch-liquidations                Monitor the liquidations feed
  bfx-cli watch-candles <market> <tf>       Monitor a candle feed
  bfx-cli watch-trades <market>             Monitor a trade feed
  bfx-cli watch-order-book [market]         Render a live order book
  bfx-cli margin-info                       Fetch and display margin information
  bfx-cli transfer <q> <ccy> <s> <d>        Transfer between wallets
  bfx-cli balances                          Fetch and display wallet balances
  bfx-cli positions                         Fetch & display all open positions
  bfx-cli order-history <market>            Fetch order history
  bfx-cli trade-history <market>            Fetch trade history
  bfx-cli close [market]                    Close open position(s)
  bfx-cli claim [market]                    Claim open position(s)
  bfx-cli order                             Submit an order
  bfx-cli currencies                        Fetch & display all available
                                            currencies
  bfx-cli tickers [market]                  Fetch and display one or multiple
                                            tickers
  bfx-cli cancel                            Cancel all open orders
  bfx-cli orders [market]                   Fetch & display all open orders
  bfx-cli status                            Query platform status
  bfx-cli download-candles <market> <tf>    Download a candle dataset
  <count>

Options:
  --version  Show version number                                       [boolean]
  --help     Show help                                                 [boolean]
```

f3rno committed:
(initial commit)

f3rno committed:
(feature) credential storage/loading, with optional encryption

f3rno committed:
(readme/refactor) update readme, remove TODO handled by API error string

f3rno committed:
(meta) update readme gif

f3rno committed:
(refactor) switch to CLI form for order submit

f3rno committed:
(feature) add bfx-hf-util catch-uncaught handler

f3rno committed:
(feature) add update-notifier

f3rno committed:
(refactor) rely on eslint for standard linting

f3rno committed:
(feature) random salt gen on install & usage

f3rno committed:
(refactor) normalize command structure

f3rno committed:
(fix) rm vestigial test

f3rno committed:
(fix) do not attempt to load config if data is piped in

f3rno commented:
@JacobPlaster please enable travis as well on this repo

f3rno committed:
(feature) add affiliate code command, rename salt file to settings

f3rno committed:
(feature) add remaining tests, set limits to reasonable levels

f3rno committed:
(manifest) bump bitfinex-api-node

f3rno committed:
(feature) add download-candles command

f3rno committed:
(fix) lower function coverage req, cannot test download command

f3rno committed:
(fix) store settings.json in module dir

eugeneglova commented on lib/commands/submit_order/process_piped.js:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  if (!args.affiliateCode) {
+    data.affiliateCode = 'xZvWHMNR'
```
default code?

f3rno replied:
Yes, it's the internal HF tracker; forgot to add configuration for a custom code, will open a new PR for it.

eugeneglova replied:
ok

f3rno committed:
(feature) add download-trades command

f3rno committed:
(fix) rm vestigial count arg from download-trades

f3rno committed:
(fix) reversed trade order, forgot to reverse unshift -> append

eugeneglova commented on README.md:
```diff
@@ -1 +1,76 @@
+bfx-cli orders [market]                   Fetch & display all open orders
+bfx-cli status                            Query platform status
+bfx-cli download-candles <market> <tf>    Download a candle dataset
+<count>
```
is it linebreak for readability ?

f3rno replied:
Autoformatted cols, imo it's best since we can reproduce later as we add commands. It's as printed in the CLI

eugeneglova commented on lib/commands/credentials/handler.js:
```diff
@@ -0,0 +1,136 @@
+
+  prompt = new Confirm({
+    name: 'encrypt-confirm',
+    message: 'Do you want to encrypt your credentials with a password? You will be prompted for the password every time you run bfx-cil:'
```
typo: `bfx-cil` should be `bfx-cli`

f3rno replied:
Fixed :+1:

eugeneglova commented on lib/commands/orders/yargs_config.js:
```diff
@@ -0,0 +1,17 @@
+  */
+module.exports = {
+  command: 'orders [market]',
+  describe: 'Fetch & display all open orders',
```
open orders? maybe placed or active?

f3rno replied:
`active` is a good idea, changed :+1:

eugeneglova commented on lib/commands/submit_order/handler.js:
```diff
@@ -0,0 +1,42 @@
+    }, true)
+
+    return script.exec(params)
+  } else if (args.length <= 1) { // piped or no args (show form)
```
`< 2` may save 1 character :-)

f3rno replied:
I held off for awhile but in the end changed it, you were right :joy: mildly easier to read/reason about

eugeneglova commented on lib/commands/submit_order/params_from_form.js:
```diff
@@ -0,0 +1,133 @@
+      parseComplete = true
+    } catch (e) {
+      console.error(colors.red(e.message))
+      console.log('')
```
empty log? just to add `\n`?

f3rno replied:
Yeah to make the output easier to read; `bitfinex-api-node` uses `debug` by default which emits a lot of noise and is not meant for data display, we override it here: https://github.com/f3rno/bfx-cli/blob/master/lib/override_bfx_api_node_debug.js

For UI we have console + `colors` module, `enquirer` for input, and `blessed-contrib` for things like OB display. Overall we need to improve the UI but things like extra newlines make a difference (80/20)

eugeneglova commented on lib/commands/submit_order/params_from_form.js:
```diff
@@ -0,0 +1,133 @@
+  // get order-specific details
+  if (
+    (params.type !== Order.type.MARKET) &&
+    (params.type !== Order.type.EXCHANGE_MARKET)
```
could be easy extended with `_.includes`
```js
!_includes([Order.type.MARKET, Order.type.EXCHANGE_MARKET], params.type)
```

f3rno replied:
Good idea, replaced below as well

eugeneglova commented on lib/commands/submit_order/params_from_form.js:
```diff
@@ -0,0 +1,133 @@
+
+  if (
+    (params.type === Order.type.STOP_LIMIT) ||
+    (params.type === Order.type.EXCHANGE_STOP_LIMIT)
```
```js
_includes([Order.type.STOP_LIMIT, Order.type.EXCHANGE_STOP_LIMIT], params.type)
```

eugeneglova commented on lib/commands/watch_candles/index.js:
```diff
@@ -0,0 +1,18 @@
+  * Watch candles command
+  *
+  * @type {Object}
+  */
```
JSDoc seems has little different paddings
```js
/**
 * 
 */
```

eugeneglova commented on lib/commands/download-trades/handler.js:
```diff
@@ -0,0 +1,63 @@
+  const start = Date.now()
+  let lastTradeMTS = +toDate
+
+  if (!+isFinite(+toDate)) {
```
`_isFinite` I see you used lodash but here native one.

f3rno replied:
Nice catch, thanks

eugeneglova commented on lib/params/order_type.js:
```diff
@@ -0,0 +1,63 @@
+      ? k.split('_')
+      : [k]
+  ))
+  .sort((a, b) => b.length - a.length)
```
https://github.com/bitfinexcom/bfx-api-node-models/blob/master/lib/order.js#L24-L28
https://github.com/bitfinexcom/bfx-api-node-models/blob/master/lib/order.js#L730
as there are keys with spaces and with underscores it's easy to use values instead
as for me this is more readable to understand
```js
_.map(
  _.reduce(Order.type, (acc, type) => {
    if (_.includes(type, 'EXCHANGE')) return acc
    return { ...acc, [type.replace('EXCHANGE ', '')]: type }
  }, {}),
  (type) => type.split(' ')
).sort((a, b) => b.length - a.length)
```

eugeneglova commented on lib/params/order_type.js:
```diff
@@ -0,0 +1,63 @@
+    if (!_isEmpty(matchedWords)) {
+      matches.push(matchedWords)
+    }
+  }
```
using `flatMap` you can save one loop on `filter` and `map` and you don't need `.push()`
```js
const matches = _.flatMap(TYPES, (type) => {
  const matchedWords = matchWord(v, type)
  return _isEmpty(matchedWords) ? [] : [matchedWords]
})
```

eugeneglova commented on lib/util/match_word.js:
```diff
@@ -0,0 +1,47 @@
+      token = tokens[j]
+
+      if (_includes(target.toLowerCase(), token)) {
+        matches.push(target)
```
is it some kind of [_.intersection](https://lodash.com/docs/4.17.15#intersection) ?

eugeneglova commented on package.json:
```diff
@@ -0,0 +1,78 @@
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/f3rno/bfx-cli.git"
```
should we use https://github.com/bitfinexcom/bfx-cli.git instead ?

f3rno replied:
Fixed, forgot about that, thanks

eugeneglova reviewed (commented):
added more suggestions/comments

f3rno committed:
(meta) add empty changelog for 1.0.0

f3rno committed:
(refactor) add log() and wrapError() utility methods

* log(...data) - passes an array of values to sprintf(), passing
  errors to wrapError() prior. Logs newline if called with no args.
* wrapError(e) - returns either the error's message or stack trace based
  on the value of the VERBOSE_ERRORS env var

f3rno committed:
(fix) rm vestigial debug() import, replaced with log()

f3rno committed:
(refactor) extract settings logic

* `/lib/db/settings.js` now loads the settings file at startup and makes
  it available as an export. Currently only used for the encryption salt
  and affilaite code

f3rno committed:
(fix) use lodash _isFinite in download-trades handler

f3rno committed:
(refactor) rename 'open orders' -> 'active orders'

(orders command)

f3rno committed:
(refactor) mild style tweak, as per @eugeneglova

f3rno committed:
(refactor) use _includes for order type-dependent handling

f3rno committed:
(fix) set custom affiliate code if possible on outgoing orders

f3rno committed:
(fix) minor indent fix

f3rno committed:
(fix) add SettingsLoadError omitted from previous commit(s)

f3rno committed:
(manifest) switch repo/bugs URLs to final bfx repo

f3rno commented:
Hi @eugeneglova thanks for this, I went back over it today. I applied most of your suggestions, but you made me rethink the order type matching logic with the questions pertaining to it. I don't like the current system, even if refactored as you suggested it would remain finicky and can cause errors due to how it parses natural language.

Do you have any suggestions for how we should parse order type strings?

Overall, my goal is to cover these sorts of cases:
* `sl` -> `stop loss`
* `s` -> `stop`
* `stop limit` -> `stop limit`
* `xl` -> `exchange limit`
* `t_s` -> `trailing stop`
* etc...

The current approach involves filtering the input to `a-Z` and replacing other chars w/ `.`, then splitting on `.` and comparing each token with each order type sub-word. It is not a very good solution, I would appreciate any suggestions you have :slightly_smiling_face: I left the relevant logic unmodified for now.

eugeneglova commented:
@f3rno not sure about that order types shortcuts, but what I noticed is you use duplicate order type in your list above `sl` and `s`. We have only `stop` type (which called stop loss usually) we don't have additional `stop loss` type.
What do you think about using this style
`l` -> `limit`
`sl` -> `stop limit`
`s` -> `stop`
`ts` -> `trailing stop`

and if we prefix these with `x` it will be `exchange`
`xl` -> `exchange limit`
`xsl` -> `exchange stop limit`
`xs` -> `exchange stop`
`xts` -> `exchange trailing stop`

f3rno commented:
@eugeneglova @ stop loss, apologies, error there was between computer & keyboard. I like your format suggestions, will implement and get back to you soon (took some time off).
